2021/CifarelliDiFranco/README.md

```diff
@@ -0,0 +1,57 @@
+# DNSLA
+A short and simple software to analyze incorrect DNS responses developed by Marco Cifarelli and Leonardo Di Franco.
+
+## Preview
+This software mainly makes use of Tshark's ring buffer capture, pyshark's (which also is Tshark but is its python wrapper) file capture and TICK stack (Telegraf, Influx, Chronograf and Kapacitor) without Telegraf. Chronograf is the general manager of the stack and can be used after its activaton at localhost on port 8888: from now on connecting Kapacitor and Influx to Chronograf will be guided by the web page. Every part of the stack can be found easily at https://docs.influxdata.com/platform/install-and-deploy/install/oss-install/.
+
+## Functioning
+As soon as the software starts, Tshark is run as a subprocess (making use of python.subprocess.Popen so that the main process doesn't have to wait its termination), with a ring buffer (Tshark's ring buffer functioning can be found at https://www.wireshark.org/docs/man-pages/tshark.html) and its pid is taken and saved for later purposes. It will continuously capture incoming traffic related to the subnet identified by the IP that's over the specified network interface and will put a .pcap file into the _livecap_ directory every 10 seconds.
+After Tshark's start, the main function will start checking _livecap_ directory's length until it is at least equal to 2 at which time it can start to extract a file in a lexicographical order every 10 seconds, analyze it with a pyshark file capture and transfer DNS relevant informations reported in the .pcap into a data structure which will be written in the influx database immediately after.  
+The data structure has this format:  
+| IPv4 | wrong dns reponses |
+| :--: | :----------------: |
+| 10.0.2.15 | 5 |
+| 10.0.2.4 | 7 |
+
+The main function also creates a shutdown_thread which will simply wait for user input until it will be equal to "quit". At that moment, it will send a INT signal to Tshark's process (he has the pid, as said before) which will stop sniffing. The main will now finish analyzing any remaning .pcap files and will terminate.
```

Comment:
Perche' c'e' bisogno di tutti questi thread che comunicano ?

Reply:
Perche' non potete analizzare il traffico nel tempo senza il bisogno di aspettare il quit ? Tutti i programmi di analisi di qcosa (dagli smartwatch al traffico di rete) funzionano cosi, ma non il vostro.

Reply:
C'è solo un thread ed è quello che resta in attesa del quit in modo da far concludere il programma in maniera elegante.
L'unico altro flusso di dati che c'è nel programma è un sotto-processo di Tshark che effettua la cattura.
Per quanto riguarda la comunicazione, il thread di shutdown si limita a mandare un semplice INT al processo di cattura quando viene rilevato un "quit" nello user input.
Non ho capito in che senso "senza il bisogno di aspettare il quit": il programma funziona sempre anche senza "quit", nel momento in cui viene dato "quit", la cattura viene interrotta, il main finisce di analizzare i pcap rimasti ed esce,

Reply:
Intendevo perche' non e' il tutto sincrono in modo da evitare queste sincronizzazioni codice complesso

Reply:
Scusi professore ma ne io ne il mio collega stiamo capendo il senso di complesso in questo caso. Il programma è un while true che analizza il traffico di rete senza mai fermarsi e non ci piaceva il fatto che per terminare dovessimo mandare un ctrl+c brutale facendo lanciare svariate eccezioni e stampe sgradevoli sul terminale. Per risolvere abbiamo implementato un semplice thread che evita questo, se è un problema lo togliamo..

Reply:
In sostanza catturate pacchetti DNS, e se rcode != 0 (che e' a un offset ben preciso del pacchetto DNS) lo contate. Una cosa che si poteva scrivere al max in 50 righe di codice senza sottoponessi e sincronizzazioni (senza parlare di tutti i problemi di sicurezza che avete innescato e di cui abbiamo discusso). Quindi che dirvi: il lavoro lo fate ma in modo molto rischioso per utente, usando disco, tenendo i privilegi di root, forkando processi e con thread. Tutto questo si poteva evitare. Solo questo. E meno codice, meno bachi meno problemi.

Reply:
Questo mi fa dispiacere, evidentemente ci siamo complicati troppo la vita.